### PR TITLE
Add support for left palette panel option

### PR DESCRIPTION
**Pull request by marthacryan**
Fixes elyra-ai/elyra#1795. Adds a prop that specifies if palette should be on the left side of the pipeline editor. 

Todo:
- [x] Fix test failures (update `palette` props in tests to be correct
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by bourdakos1 — APPROVED** [review]
lgtm

**Comment by marthacryan** [comment]
Found an issue with this PR - going to leave in draft until it's fixed up

File: packages/pipeline-editor/src/PipelineEditor/index.tsx [hunk 1]
```diff
@@ -482,7 +484,7 @@ const PipelineEditor = forwardRef(
 
         if (e.editType === "toggleOpenPanel") {
           if (!panelOpen) {
-            let defaultTab = "palette";
+            let defaultTab = leftPalette ? "pipeline-properties" : "palette";
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineEditor/index.tsx:487]: 👀 lint

**Review by ajbozarth — CHANGES_REQUESTED** [review]
Code LGTM, tested it locally and over looks solid, but the side bar UI is wonky when closed, as seen in this screenshot the text is cut off, we should either replace that text with an icon or figure something else out. 
Screen Shot 2021-06-24 at 2 46 41 PM

**Comment by ajbozarth** [comment]
also personally I think as long as we only have one tab on the left for pallete we should have it default to open, at minimum we should have it default to open for an empty pipeline

**Comment by akchinSTC** [comment]
@lresende - when do we think we can get this in by? early next week? Just need to finalize the look of the UI and flow so that ptitzler can work on taking screens for documentation etc.

**Comment by lresende** [comment]
Can we get a screenshot of the current status? If the above screenshot is current we need to make it better, maybe by using component icon + tooltip? We should have something very similar to the one on the issue.

**Comment by ptitzler** [comment]
Feedback based on today's build.

- There appears to be a palette issue (the entires only show up after I randomly click on things).
  
   empty

- The text should label these as _component[s]_ instead of _node[s]_ due to new terminology we are introducing (see https://github.com/elyra-ai/elyra/pull/1839)  In a nutshell, a [pipeline] node is an  instance of a generic or typed component. This way our terminology is consistent with KF.
image


image

- Perhaps use a dark font color? As is there is too little contrast between white and gray. The color choices should probably be revisited for the dark theme (:
- Does the "nodes" text serve a purpose?
- What's the source for the 2 component specific labels for the three generic components? It appears that for other components the second label is the description. If that's the case, the descriptions for the three generic components should probably be "Run <Python script| notebook | R script> 


image

- Is there a way to make the collapsed view more useful to the user? As is one would have to expand the view or hover over each custom component entry to determine which one is which. Going forward the majority of entries in the list will be custom components and the icon the same for all, which gives the current rendering a "buggy" appearance. If there's no good way I'd +1 Alex's suggestion to show the component  list in the palette expanded by default.
- For the time being, could the list be sorted? Right now the order appears to reflect the order in the registry, which is random.

**Comment by marthacryan** [comment]
@lresende @ptitzler The "Nodes" dropdown is because of the setup of the palette - it requires using categories, so we have just one category called nodes (which was coded to not be shown in the current implementation of the palette as a tab). Unfortunately it doesn't seem to be possible to just give a list of nodes without any categories (or have a category selected by default), but we have a couple options:
* Add more categories so that it looks more natural to have categories
* Reuse the `PalettePanel` - the one part about this would be that it might require a good bit of extra work to have a "Narrow" mode like the built-in palette has. If we don't care about the "narrow" mode or if we can put that off a bit, I think this might be the best solution

**Comment by ptitzler** [comment]
> Add more categories (with icons?) so that it looks more natural to have categories

Have we tried yet to define what what categories are? The most simplistic approach would distinguish between generic components ("Elyra components that are supported by every runtime) and custom components (the others), but won't help  in the mid- or long-term because the majority of entries would always be in the second category.

> Reuse the PalettePanel class currently used in the right panel - the one part about this would be that it might require a good bit of extra work to have a "Narrow" mode like the built-in palette has. It also doesn't have a search feature yet. If we don't care about the "narrow" mode or if we can put that off a bit, I think this might be the best solution

I feel that "we don't care about the "narrow" mode" -> defer might be the right thing to do initially:
- since the palette is new and users not familiar with the example components, it's likely that users would choose the expanded view because it helps them decide which components to use
- we could use the button to show/hide the palette to avoid using up precious canvas space permanently until a more useful narrow mode is available.
- as is the narrow mode clashes a bit with category labels (which I guess would be text based) because the narrow mode size would have to be at least as wide as the longest category name

**Comment by marthacryan** [comment]
Updated to take entire `palette` as a prop, rather than `nodes`, which allows the JL extension to define categories / icons for categories as needed.

**Comment by ptitzler** [comment]
Layout wise this seems a lot better with the latest updates! Search etc looks fine!

- I'm still seeing the font/theme color issues
- Nit: `Palette` tooltip ->  `Minimize Palette/Maximize Palette"  (similar to what we have for the properties panel)
image
- Could you please add a default description for each generic node (`Run XX`)?
  image

**Comment by ptitzler** [comment]
Might want to update the wording here to help new users get started (look for the palette!)

image

**Comment by ptitzler** [comment]
I didn't test this with "many" component entries, but assume that the palette will have its own scroll bar?

File: packages/pipeline-editor/src/PipelineController/index.test.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ import {
   InvalidPipelineError,
   ElyraOutOfDateError,
 } from "../errors";
-import { nodeSpec } from "../test-utils";
+import { createPalette, nodeSpec } from "../test-utils";
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.test.ts:22]: haha smart!

File: packages/pipeline-editor/src/PipelineController/index.ts [hunk 3]
```diff
@@ -430,6 +434,13 @@ class PipelineController extends CanvasController {
     return nodes;
   }
 
+  getAllPaletteNodes(): any[] {
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.ts:437]: we should specify the actual type here

File: packages/pipeline-editor/src/PipelineController/index.ts [hunk 4]
```diff
@@ -455,11 +466,11 @@ class PipelineController extends CanvasController {
 
     if (node !== undefined) {
       const { op, app_data } = node;
-      const nodeDef = this.nodes.find((n) => n.op === op);
+      const nodeDef = this.getAllPaletteNodes().find((n) => n.op === op);
 
       const info = nodeDef?.properties?.uihints?.parameter_info ?? [];
 
-      const properties = info.map((i) => {
+      const properties = info.map((i: any) => {
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.ts:473]: once we define a real type in `getAllPaletteNodes` we should be able to remove `any` here

**Review by bourdakos1 — CHANGES_REQUESTED** [review]
looks great! Just a request to add better typings for one of the new functions

**Comment by marthacryan** [comment]
> I didn't test this with "many" component entries, but assume that the palette will have its own scroll bar?

@ptitzler Just checked and it does

> * I'm still seeing the font/theme color issues
Fixed!
image
image
image
image

> * Nit: `Palette` tooltip ->  `Minimize Palette/Maximize Palette"  (similar to what we have for the properties panel)
>   image
This is actually not configurable, it's built in to the canvas. 

> * Could you please add a default description for each generic node (`Run XX`)?
>   image
That'll require backend changes on the Elyra side. @kiersten-stokes Maybe you could add more descriptions to the nodes?

**Comment by marthacryan** [comment]
@bourdakos1 How does that last commit look for types?

File: packages/pipeline-editor/src/PipelineController/index.ts [hunk 5]
```diff
@@ -430,6 +435,13 @@ class PipelineController extends CanvasController {
     return nodes;
   }
 
+  getAllPaletteNodes(): CustomNodeSpecification[] {
+    return (
+      this.palette?.categories?.flatMap((cat: CategoryDef) => cat.node_types) ??
+      ([] as any[])
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.ts:441]: we shouldn't need to cast this as `any[]`

File: packages/pipeline-editor/src/PipelineController/index.ts [hunk 6]
```diff
@@ -396,12 +394,14 @@ class PipelineController extends CanvasController {
...
+          const nodeDef = this.getAllPaletteNodes().find(
+            (n) => n.op === node.op
+          );
 
           const message = nodeProblems
             .map((problem) => {
-              const label = nodeDef.properties!.uihints!.parameter_info.find(
-                (p) => p.parameter_ref === problem.property
+              const label = nodeDef!.properties!.uihints!.parameter_info.find(
+                (p: any) => p.parameter_ref === problem.property
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.ts:404]: no `any`

File: packages/pipeline-editor/src/test-utils.tsx [hunk 7]
```diff
@@ -322,6 +324,68 @@ export const samplePipeline = {
...
+      app_data: {
+        ui_data: {
+          label: node.label,
+          description: node.description,
+          image: node.image ?? "",
+          x_pos: 0,
+          y_pos: 0,
+        },
+      },
+      ...node,
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/test-utils.tsx:383]: I think this is what was causing your typescript issues. `node` has `type` which would override `type: "execution_node",` above

**Review by bourdakos1 — CHANGES_REQUESTED** [review]

File: packages/pipeline-editor/src/types.ts [hunk 8]
```diff
@@ -26,6 +26,12 @@ export interface CustomNodeSpecification {
   extension?: string;
   image?: string;
   properties?: PropertyDefinitions;
+  id?: string;
+  type?: "execution_node";
+  inputs?: any[];
+  outputs?: any[];
+  parameters?: any;
+  app_data?: any;
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/types.ts:34]: I think we should remove `CustomNodeSpecification` from here

File: packages/pipeline-editor/src/PipelineController/index.ts [hunk 9]
```diff
@@ -430,6 +435,12 @@ class PipelineController extends CanvasController {
     return nodes;
   }
 
+  getAllPaletteNodes(): ExecutionNodeDef[] {
+    return (this.palette?.categories?.flatMap(
+      (cat: CategoryDef) => cat.node_types
+    ) ?? []) as ExecutionNodeDef[];
+  }
+
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/PipelineController/index.ts:443]: I need to take a closer look at this, seems a bit fishy

File: packages/pipeline-editor/src/common-canvas.d.ts [hunk 10]
```diff
@@ -710,6 +710,37 @@ declare module "@elyra/canvas" {
      * Operator type identifier
      */
     op: string;
+
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/common-canvas.d.ts:713]: need to take a closer look at this too

**Review by bourdakos1 — APPROVED** [review]

File: packages/pipeline-editor/src/common-canvas.d.ts [hunk 11]
```diff
@@ -2956,2314 +2683,2613 @@ declare module "@elyra/canvas" {
...
+      runtime_ref?: string;
+      /**
+       * Object containing app-specific data
+       */
+      app_data: {
+        // NOTE: app_data shouldn't but undefined in elyra.
+        //====================================================================//
+        // Elyra specific fields
+        //====================================================================//
+        /**
```
Comment (bourdakos1) [inline packages/pipeline-editor/src/common-canvas.d.ts:2753]: Label and  image should be removed from here

**Review by bourdakos1 — CHANGES_REQUESTED** [review]

**Review by bourdakos1 — APPROVED** [review]

**Review by ajbozarth — APPROVED** [review]
skimmed the code and checked the UI/UX locally and it looks ok, trusting other for the more through review

**Pull request merged**